hive external table add drop ename column

Zhangruichao opened the pull request:
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #4030

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

Zhangruichao committed:
async hive add rename or drop column

Zhangruichao committed:
sync hive add rename or drop column

Zhangruichao committed:
sync hive add rename or drop column

Zhangruichao committed:
persistence ModifyTableColumnOperationLog

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=4319) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=4319) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=4319) it.</sub>

wuqiao commented:
@Zhangruichao hi ruichao, Please add the StarRocks assistant  WeChat:"StarRocks-1". We can communicate  in group.

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:
```diff
@@ -180,32 +180,7 @@
-import com.starrocks.persist.TableInfo;
-import com.starrocks.persist.TablePropertyInfo;
-import com.starrocks.persist.TruncateTableInfo;
+import com.starrocks.persist.*;
```
Don't use '*' in here. you could flat it.

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:
```diff
@@ -202,4 +202,6 @@
     // workgroup 10021 ~ 10030
     public static final short OP_WORKGROUP = 10021;
+
+    public static final short OP_MODIFY_HIVE_TABLE_COLUMN = 10031;
```
plz add some comment

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -37,18 +37,16 @@
-import com.starrocks.external.hive.HivePartitionStats;
-import com.starrocks.external.hive.HiveTableStats;
-import com.starrocks.external.hive.Utils;
+import com.starrocks.external.hive.*;
```
Ditto.

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +187,55 @@ public HiveTableStats getTableStats() throws DdlException {
-                .refreshTableCache(resourceName, hiveDb, hiveTable, getPartitionColumns(),
-                        new ArrayList<>(nameToColumn.keySet()));
+    public void refreshTableCache(String dbName, String tableName) throws DdlException {
+        org.apache.hadoop.hive.metastore.api.Table table = Catalog.getCurrentCatalog().getHiveRepository()
```
import org.apache.hadoop.hive.metastore.api.Table;

Zhangruichao replied:
org.apache.hadoop.hive.metastore.api.Table;  and com.starrocks.catalog.Table; cannot exist at the same time；

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +187,55 @@ public HiveTableStats getTableStats() throws DdlException {
+        for (FieldSchema hiveColumn : partHiveColumns) {
+            allHiveColumns.put(hiveColumn.getName(), hiveColumn);
+        }
+        boolean needRefreshColumn = allHiveColumns.size() != this.fullSchema.size();
```
I think there are some tricks to judging whether a table has schema changes by the number of columns. If a user adds a column and drop a column, Both are equal. So I think you may judge if each column is the same.

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/persist/ModifyTableColumnOperationLog.java:
```diff
@@ -0,0 +1,71 @@
+// This file is made available under Elastic License 2.0.
```
Please use Elastic License

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +187,55 @@ public HiveTableStats getTableStats() throws DdlException {
+        boolean needRefreshColumn = allHiveColumns.size() != this.fullSchema.size();
+        if (!needRefreshColumn) {
+            for (Column column : fullSchema) {
+                FieldSchema fieldSchema = allHiveColumns.get(column.getName());
```
I think it can't guarantee that the two columns are exactly the same through judging only by the column name

Zhangruichao committed:
through column type and name judge different

Zhangruichao committed:
com.starrocks.persist.*; import all;
ModifyTableColumnOperationLog under Elastic license;
add OperationType comment;

Zhangruichao committed:
import com.starrocks.external.hive.*; fix

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +191,57 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
+        List<FieldSchema> unPartHiveColumns = table.getSd().getCols();
+        List<FieldSchema> partHiveColumns = table.getPartitionKeys();
+        Map<String, FieldSchema> allHiveColumns = unPartHiveColumns.stream()
+                .collect(Collectors.toMap(FieldSchema::getName, fieldSchema -> fieldSchema));
```
Please write this part in a method

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +191,57 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
+        for (FieldSchema hiveColumn : partHiveColumns) {
+            allHiveColumns.put(hiveColumn.getName(), hiveColumn);
+        }
+        boolean needRefreshColumn = allHiveColumns.size() != this.fullSchema.size();
```
```suggestion
        boolean columnsSizeEqual = allHiveColumns.size() != this.fullSchema.size();
```

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +191,57 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
+                    break;
+                }
+                PrimitiveType type = toPrimitiveType(fieldSchema.getType());
+                if (!type.name().equalsIgnoreCase(column.getType().toString())) {
```
Please write this part in a method

stephen-shelby replied:
```suggestion
                if (type != column.getType().getPrimitiveType()) {
```

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +191,57 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
+                    .refreshTableCache(resourceName, hiveDb, hiveTable, getPartitionColumns(),
+                            new ArrayList<>(nameToColumn.keySet()));
+        } else {
+            updateFullSchema(dbName, tableName, allHiveColumns);
```
record edit log should be place at the end

stephen-shelby replied:
If modify full schema, the column stats cache should be cleaned up

stephen-shelby replied:
schema changes should also be adapted in MetastoreEventProcessor

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/journal/JournalEntity.java:
```diff
@@ -52,42 +52,7 @@
-import com.starrocks.persist.TablePropertyInfo;
-import com.starrocks.persist.TruncateTableInfo;
-import com.starrocks.persist.WorkGroupOpEntry;
+import com.starrocks.persist.*;
```
flat *

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +191,57 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
+        for (Map.Entry<String, FieldSchema> entry : allHiveColumns.entrySet()) {
+            FieldSchema fieldSchema = entry.getValue();
+            PrimitiveType type = toPrimitiveType(fieldSchema.getType());
+            Column column = new Column(fieldSchema.getName(), ScalarType.createType(type), true, null, true,
```
isKey、aggType、defaultValue couldn't hard code

Zhangruichao committed:
code structure optimization

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java:
```diff
@@ -957,7 +963,8 @@ public void logRecoverPartition(RecoverInfo info) {
 
-    public void logModifyPartition(ModifyPartitionInfo info) {
+    public void
+    logModifyPartition(ModifyPartitionInfo info) {
```
delete enter

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/catalog/HiveTable.java:
```diff
@@ -188,10 +190,69 @@ public Map<String, HiveColumnStats> getTableLevelColumnStats(List<String> column
+        }
+    }
+
+    private void refreshTableCacheByColumnNames(List<String> columnNames) throws DdlException {
```
```suggestion
    private void refreshTableCache(List<String> columnNames) throws DdlException {
```

Zhangruichao committed:
code structure optimization

Zhangruichao committed:
code structure optimization

Zhangruichao committed:
add UT

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java:
```diff
@@ -809,6 +809,12 @@ public static void loadJournal(Catalog catalog, JournalEntity journal) {
                     catalog.getAnalyzeManager().replayRemoveAnalyzeJob(analyzeJob);
                     break;
                 }
+                case OperationType.OP_MODIFY_HIVE_TABLE_COLUMN:{
```
If this log is only for hive table, please change the catalog.replayModifyTableColumn to catalog.replayModifyHiveTableColumn. If this is for all kind table, change OP_MODIFY_HIVE_TABLE_COLUMN to OP_MODIFY_TABLE_COLUMN

Zhangruichao committed:
rename replayModifyHiveTableColumn
